app/src/main/java/com/eveningoutpost/dexdrip/Services/LibreWifiReader.java

```diff
@@ -0,0 +1,620 @@
+            recieversIpAddresses = "fake://FAKE_DATA";
+        } else {
+            recieversIpAddresses = Pref.getString("wifi_recievers_addresses", "");
+        }
+        int packetsToRead = 1;
+        Log.d(TAG, "reading " + packetsToRead + " packets");
+        LibreWifiData[] LibreWifiDataArr = Read(recieversIpAddresses, packetsToRead);
+        
+        Log.d(TAG, "After reading ..." + LibreWifiDataArr);
+
+        if (LibreWifiDataArr == null || LibreWifiDataArr.length == 0) {
+            return;
+        }
+        Log.d(TAG, "After verification ..." + LibreWifiDataArr);
+        // Last in the array is the most updated reading we have.
+        for (LibreWifiData LastReading : LibreWifiDataArr) {
+            // Last in the array is the most updated reading we have.
+            //TransmitterRawData LastReading = LastReadingArr[LastReadingArr.length -1];
+
+
+            //if (LastReading.CaptureDateTime > LastReportedReading + 5000) {
+            // Make sure we do not report packets from the far future...
+            if ((LastReading.CaptureDateTime > LastReportedTime + 4 * 60000) &&
+                    LastReading.CaptureDateTime < new Date().getTime() + 120000) {
+                // We have a real new reading...
+                Log.d(TAG, "calling HandleGoodReading from " +  JoH.dateTimeText(LastReading.CaptureDateTime ));
+
+                byte data[] = Base64.decode(LastReading.BlockBytes, Base64.DEFAULT);
+                boolean checksum_ok = NFCReaderX.HandleGoodReading("tomato", data, LastReading.CaptureDateTime);
+                if (checksum_ok) {
+                    // TODO use battery, and other interesting data.
+                    LastReportedTime = LastReading.CaptureDateTime;
+                } else {
+                    Log.e(TAG, "Recieved a pacjet with bad checksum");
+                }
+            }
+        }
+    }
+    
+
+    // data for MegaStatus
+    static List<StatusItem> megaStatus() {
+        final List<StatusItem> l = new ArrayList<>();
+        for (Map.Entry<String, String> entry : hostStatus.entrySet()) {
+            final long status_time = hostStatusTime.get(entry.getKey());
+            if (entry.getValue().length() > 0)
+                l.add(new StatusItem(entry.getKey(), entry.getValue() + ((status_time != 0) ? (" " + JoH.niceTimeSince(status_time) + " " + "ago") : ""), JoH.msSince(status_time) <= BgGraphBuilder.DEXCOM_PERIOD ? StatusItem.Highlight.GOOD : JoH.msSince(status_time) <= BgGraphBuilder.DEXCOM_PERIOD * 2 ? StatusItem.Highlight.NOTICE : StatusItem.Highlight.NORMAL));
+        }
+        return l;
+    }
+
+    static void statusLog(String key, String msg) {
+        statusLog(key, msg, 0); // default no time since
+    }
+
+    // timestamp or 0 = don't use or -1 = now
+    static void statusLog(String key, String msg, long time) {
+        hostStatus.put(key, msg);
+        hostStatusTime.put(key, (time != -1) ? time : JoH.tsl());
+    }
+}
```

Comment:
This looks like a significant duplication of code. I don't propose any changes at this point, but what are the constraints that prevent us for reusing the existing wifi-reader for this purpose?

Reply:
So, the main reason that we were not able to reuse the existing code are:
1) The need to move to a new protocol (see in the main answer why).
2) The existing code was returning a list of <TransmitterRawData>. Since libre has other fields (see LibreWifiData), one would have to have a third class that both will inherit from. this will make code more complex.
3) The way multiple points are handled: When reading from the g4 each packet matters. when reading from libre, we can skip older readings. That said, if for example last packet contains something like "no sensor" we need to move to the previous packet. So, libre code has to be different.